[pipelining] Add schedule runtime for lowered schedule

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #131762
* #130860
* __->__ #130488

Creates a new runtime that shifts complexity from runtime to
ahead-of-time.

The existing runtime (PipelineScheduleMulti) accepts a
compute-only schedule (forward, backward, weight) actions only are
specified, and it infers the communication operations at runtime.
Compared to that runtime, PipelineScheduleRuntime has less logic that
happens at runtime and relies on lowering passes to transform the
compute-only schedule to add communications.

Advantages include
- easier to verify the correctness by dumping a compute+comm schedule
- posible to manually edit the compute+comm schedule if the lowering
  heuristics are insufficient

Functionality included inside the PipelineScheduleRuntime is limited to
- accepting a compute-only schedule and lowering it to add comms
- executing the compute or comm operations specified by the given
  schedule
- handling work.wait() automatically by calling it just before the
  matching compute operation (for RECV ops) or at the end of step (for
  SEND ops)

Follow ups for later PRs
- Some refactoring should be done to replace PipelineScheduleMulti with
  this runtime
- Optimizer execution is not considered (e.g. for zero-bubble cases)

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @d4l3k @c-p-i-o

wconstab committed:
Update

[ghstack-poisoned]

wconstab committed:
Update

[ghstack-poisoned]

wconstab committed:
Update

[ghstack-poisoned]

wconstab committed:
Update

[ghstack-poisoned]

wconstab committed:
Update

[ghstack-poisoned]

wconstab committed:
Update

[ghstack-poisoned]

wconstab committed:
Update

[ghstack-poisoned]

H-Huang reviewed (approved):
Overall looks good to me, there are a lot of asserts in the lowered schedule runtime, so potentially could be cleaner if we do validation on the schedule itself.

wconstab committed:
Update

[ghstack-poisoned]

wconstab commented on torch/distributed/pipelining/schedules.py:
```diff
@@ -1257,6 +1272,254 @@ def _step_microbatches(
+    def _load_actions(
+        self,
+        actions: Dict[int, List[Optional[_Action]]],
+        format: str = "compute_only",
```
i don't think so.

compute_comms is not a very well defined format per-se.  It really just means "the schedule has everything in it, don't do any passes".

there is some optionality of which passes to run though.  (or maybe which modes to run them in).  For example batching or merging BW are optional,  but adding send/recv is required, and unshard/reshard is required too (if using FSDP)

wconstab commented on torch/distributed/pipelining/schedules.py:
```diff
@@ -837,13 +851,13 @@ def _reshard(stage_index: int):
-            fetch,
-            evict,
-        )
+        # logger.debug(
```
non-accidental. i felt it was too verbose even for when i enable + logging and want to look at other verbose logs.  But i felt like its worth leaving in in case i was ever debugging this pass again.

wconstab committed:
Update

[ghstack-poisoned]

wconstab committed:
Update

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge -i

wconstab commented:
@pytorchbot merge -i

PaliC commented:
@pytorchbot revert -m "breaking distributed tests internally (that should be running in OSS)" -c "ghfirst"

Discussed this offline with Will

wconstab committed:
Update

[ghstack-poisoned]

wconstab committed:
Update

[ghstack-poisoned]

wconstab committed:
Update on "[pipelining] Add schedule runtime for lowered schedule"


Creates a new runtime that shifts complexity from runtime to
ahead-of-time.

The existing runtime (PipelineScheduleMulti) accepts a
compute-only schedule (forward, backward, weight) actions only are
specified, and it infers the communication operations at runtime.
Compared to that runtime, PipelineScheduleRuntime has less logic that
happens at runtime and relies on lowering passes to transform the
compute-only schedule to add communications.

Advantages include
- easier to verify the correctness by dumping a compute+comm schedule
- posible to manually edit the compute+comm schedule if the lowering
  heuristics are insufficient

Functionality included inside the PipelineScheduleRuntime is limited to
- accepting a compute-only schedule and lowering it to add comms
- executing the compute or comm operations specified by the given
  schedule
- handling work.wait() automatically by calling it just before the
  matching compute operation (for RECV ops) or at the end of step (for
  SEND ops)

Follow ups for later PRs
- Some refactoring should be done to replace PipelineScheduleMulti with
  this runtime
- Optimizer execution is not considered (e.g. for zero-bubble cases)

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 d4l3k c-p-i-o

[ghstack-poisoned]

wconstab committed:
Update

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge -i

wconstab commented:
@pytorchbot merge
